mono/metadata/mono-perfcounters.c

```diff
@@ -1483,276 +1567,438 @@ typedef struct {
+		goto_if_nok (error, Exit);
+	}
+	goto Exit;
+	
+Null:
+	MONO_HANDLE_SET_NULL (MonoArray, result);
+	goto Exit;
+	
+Empty: // Return a zero sized array, not null, no error.
+	result = mono_array_new_handle (domain, mono_get_string_class (), 0, error);
+	goto Exit;
 
-		for (i = 0; i < scat->num_counters; ++i) {
-			MonoString *str = mono_string_new_checked (domain, p + 1, &error);
-			goto_if_nok (&error, leave);
-			mono_array_setref (res, i, str);
-			p += 2; /* skip counter type */
-			p += strlen (p) + 1; /* skip counter name */
-			p += strlen (p) + 1; /* skip counter help */
-		}
-	} else
-		res = mono_array_new_checked (domain, mono_get_string_class (), 0, &error);
-leave:
-	perfctr_unlock ();
-	mono_error_set_pending_exception (&error);
-	return res;
+Exit:
+	if (unlock)
+		perfctr_unlock ();
+	mono_unwrapped_string_cleanup (&category);
+	return result;
+}
+
+static void
+get_string_array_loop (MonoArrayHandle result, void **array, int i, gboolean is_process, MonoError *error)
+// Avoid creating co-op handles in loops.
+{
+	HANDLE_FUNCTION_ENTER()
+
+	char buf [128];
+	char *p;
+	if (is_process) {
+		char *pname = mono_process_get_name (array [i], buf, sizeof (buf));
+		p = g_strdup_printf ("%d/%s", GPOINTER_TO_INT (array [i]), pname);
+	} else {
+		sprintf (buf, "%d", GPOINTER_TO_INT (array [i]));
+		p = buf;
+	}
+	MonoStringHandle str = mono_string_new_handle (mono_domain_get (), p, error);
+	if (p != buf)
+		g_free (p);
+	goto_if_nok (error, Exit);
+	MONO_HANDLE_ARRAY_SETREF (result, i, str);
+Exit:
+	HANDLE_FUNCTION_RETURN ();
 }
 
-static MonoArray*
+static MonoArrayHandle
 get_string_array (void **array, int count, gboolean is_process, MonoError *error)
 {
-	int i;
-	MonoDomain *domain = mono_domain_get ();
-	error_init (error);
-	MonoArray * res = mono_array_new_checked (mono_domain_get (), mono_get_string_class (), count, error);
-	return_val_if_nok (error, NULL);
-	for (i = 0; i < count; ++i) {
-		char buf [128];
-		char *p;
-		if (is_process) {
-			char *pname = mono_process_get_name (array [i], buf, sizeof (buf));
-			p = g_strdup_printf ("%d/%s", GPOINTER_TO_INT (array [i]), pname);
-		} else {
-			sprintf (buf, "%d", GPOINTER_TO_INT (array [i]));
-			p = buf;
-		}
-		MonoString *str = mono_string_new_checked (domain, p, error);
-		if (p != buf)
-			g_free (p);
-		return_val_if_nok (error, NULL);
-		mono_array_setref (res, i, str);
+	HANDLE_FUNCTION_ENTER()
+
+	MonoArrayHandle result = mono_array_new_handle (mono_domain_get (), mono_get_string_class (), count, error);
+	goto_if_nok (error, Null);
+
+	for (int i = 0; i < count; ++i) {
+		get_string_array_loop (result, array, i, is_process, error);
+		goto_if_nok (error, Null);
 	}
-	return res;
+
+	goto Exit;
+
+Null:
+	MONO_HANDLE_SET_NULL (MonoArray, result);
+Exit:
+	HANDLE_FUNCTION_RETURN_OBJ (result);
 }
 
-static MonoArray*
+static void
+get_string_array_of_strings_loop (MonoArrayHandle result, void **array, int i, MonoError *error)
```

Comment:
I wonder if longer-term we couldn't find some way to move all this array of strings manipulation up to managed code. We do have the ability to safely work with, for example, a `GPtrArray` of UTF-8 c strings from managed code (see https://github.com/mono/mono/tree/master/mcs/class/corlib/Mono - `RuntimeMarshal.cs` and `SafeGPtrArrayHandle.cs`)

Reply:
Agreed but didn't tackle it today. CoreCLR appears to have written/rewritten perfcounter in C#.